### ISSUE TITLE
Classifying with Checkpoint

### ISSUE THREAD

**Issue by Colhodm** [kratzert/finetune_alexnet_with_tensorflow#38]
Hey, so I finished fine tuning my model, but I haven't really used tensorflow that much before. I understand how to load in the images I want to test on, but how do I use the checkpoint file to classify them?

Also thoughts on parameters for this dataset; 330 training images of Benign and Malignant, along with a validation and testing set of 40 images. I changed learning rate to 0.005 as well as batch size to 40, any other thoughts? I'm still training the last two layers, I'm not sure whether this is the right move.
Thanks!!

**Comment by Colhodm** [comment]
Hi so I've actually almost finished this process, I just have one last issue, which is that the dimensions of the images are 227 by 227,so how can you subtract the imagenet mean from them with dimensions 1 by 3?

**Comment by kratzert** [comment]
You can load weights from checkpoints using the tf.Saver().restore() and subtraction the imagenet mean is done 2 times in the code of the datagenerator.py. 
You could have just looked it up there.

img_centered = tf.subtract(img_resized, IMAGENET_MEAN)

**Comment by Colhodm** [comment]
Hi,
I almost finished the classifying script: but I'm getting the same output for probabilities for every training the image, and the dimensions of 316 don't make sense. Could you explain what I'm doing wrong in this code? 
Here is the file: https://github.com/Colhodm/TensorFlow-Classifier-Alex_Net/blob/master/test.py
Thanks,
Arjun

**Comment by kratzert** [comment]
Hi Arjun,

I had a look at your file but it is really messy, with a lot of uncommented code lines, prints etc that it is hard to follow your actual code. Anyway I noticed you are using tensorflows session wrong. What you actually have to do is everything you want to do in one session (like restoring weights and then classifiying) has to be in the scope of the with tf.Session() as sess: block

like:

    # do all the setup here
   
    with tf.Session() as sess:
        # Here comes the restoring and the classification

So I'm actually unsure if you even import the weights to the correct session, since you do it outside of the block. Anyway, I'm not sure this is the reason for your problem, but it is not correct usage of TF anyway

**Comment by Colhodm** [comment]
Hi, 
Could you look through it again; I cleaned out the comments and improved the readability; its linked to the same repo: https://github.com/Colhodm/TensorFlow-Classifier-Alex_Net/blob/master/test.py
The couple of questions I had were that; first I didn't understand why the dimensions of prob for me were (316 by 2), and how to classify using softmax. My test set is 40 images, which is why I was confused by the dimensions for prob. I trained the last two layers of the CNN with number of classes being 2 (Benign and Malignant). Everything should be restored and classified within the tensorflow session, so that shouldn't be an issue.
Thanks for all the time, will cite your code in the paper!
Arjun

**Comment by kratzert** [comment]
Hard to say without seeing your graph. How many classes do you have (or how many neurons you have in the last fc8 layer?) 

`prob` should have the dimension of 1 times the neurons you have in your last layer in each loop.

Another general note to finetuning: During training and then later during inference, just subtract the imagenet mean, not your training set mean as well. You want the conv layers to work as feature extractor as they were trained on the original data set. If you now subtract another mean, they are in completely different scales and the entire conv filters might not work anymore. But this has nothing to do with your problem, just a general note.

**Comment by Colhodm** [comment]
The parameters I specified were a batch size of 20,learning rate of 0.005 and num_classes = 2. The training set of images was 326 images. How did you intend on people loading in the weights for classifying from the checkpoint? I assumed it was by loading   score  = sess.run('fc8/weights:0') from the checkpoints. The len(score) 4096 and the len of probs is also (4096) so they match earlier like you said they should. I'm just confused about this line:  probs = sess.run(softmax, feed_dict={x: img, keep_prob: 1}) and how to actually get predictions from it. I understand it would be through argmax, but I don't follow the dimensions, since my test set is 40 images.

If you have a sample script of classifying images with a fine tuned net loaded from a checkpoint, this could be more helpful, but right now I'm looking at your validate.ipynb for general tips on how to go about it.

Sorry for all the questions!
Thanks so much!
Arjun

**Comment by kratzert** [comment]
Your problem is, that you restore the graph from the training, not only the weights. Just do create the graph from my AlexNet class like in the training script, like the fc8 output to tf.nn.softmax and then within the tf.Session only restore the weights.

Should look like this

```python
# initialize input place holder to specific batch_size, e.g. 1 if you want to classify image by image
x = tf.placeholder(tf.float32, [batch_size, 227, 227, 3])
keep_prob = tf.constant(1., dtype=tf.float32)

# Initialize model
model = AlexNet(x, keep_prob, num_classes, train_layers)

# Link variable to model output
score = model.fc8
softmax = tf.nn.softmax(score)

# create saver instance
saver = tf.train.Saver()

with tf.Session() as sess:    
    self.session.run(tf.global_variables_initializer())
    saver.restore(self.session, 'path/to/*.ckpt')

    # Now comes your image loop
    for img in imgs:
         pred = sess.run(score, feed_dict={x: img})

         predicted_label = pred.argmax(axis=1)

```
Sometime like this could look your code. If this still doesn't help, I would recommend you read some more tutorials about TensorFlow or Python.

**Comment by Colhodm** [comment]
Hi Frederik,
Thanks for all the help; I really appreciate it, I got the classifier to work!
Thanks,
Arjun

**Comment by Colhodm** [comment]
Hi Frederik,
I have a couple of questions: first the training always seems to diverge when the learning rate is greater than 0.001(nan's in histogram), I was wondering whether you had any insight into that. The data is normalized so I'm not sure as to why this is happening. Also I finished the classifier, but I just wanted to make sure that all my image manipulations, e.g reversing the image at the end before calculating probabilities all made sense. The code is in this repo: https://github.com/Colhodm/TensorFlow-Classifier-Alex_Net/blob/master/classifer.py
If you could quickly look through it and see if there are any red flags that would be really helpful.
Thanks so much!
Arjun
